Toggl.Networking/Network/ApiClients/AdaptingRateLimitingAwareApiClient.cs

```
@@ -0,0 +1,79 @@
+        }
+
+        public void Dispose()
+        {
+            internalClient.Dispose();
+        }
+
+        public async Task<IResponse> Send(IRequest request)
+        {
+            var repeat = false;
+            var response = await internalClient.Send(request);
+
+            if (response.StatusCode == HttpStatusCode.TooManyRequests)
+            {
+                repeat = !burstsDisabled;
+                burstsDisabled = true;
+            }
+
+            if (response.TryGetBucketSizeFromHeaders(out var size))
+            {
+                // We assume that under normal circumstances (there is no unexpected heavy load on the server) the burst
+                // size is at the normal size (10). When the bucket size is decreased dynamically and is below a threshold (80),
+                // we will assume that the burst size has dropped to zero as well. This isn't guaranteed by the ops team
+                // though and it might be possible that the burst size is lower even when the bucket size drops.
+                //
+                // If we get a "Too Many Requests" error, we will stop sending requests in bursts until the app restarts
+                // or until the size of the bucket increases.
+
+                lock (leakyBucket)
+                {
+                    if (burstsDisabled && size > leakyBucket.Size && response.StatusCode != HttpStatusCode.TooManyRequests)
+                    {
+                        burstsDisabled = false;
+                    }
+
+                    var burstSize = size >= normalBucketSize && !burstsDisabled ? assumedBurstSize : noBursts;
+                    leakyBucket.ChangeSize(size, burstSize);
+                }
+            }
+
+            if (repeat)
+            {
+                // let the server bucket leak for a while
+                await Task.Delay(assumedBurstSize * leakyBucket.SlotDuration);
+
+                return await Send(request);
+            }
```

This is complicated and potentially swallows a lot of 429s that we want to track with analytics. I think it would be safer to just fail fast instead. We don't get a lot of 429s right now, and we can keep monitoring it to see how it goes and always make improvements later. :)